[WIP] Add histograms to calibration plots

aperezlebel opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Will close #23709 

#### What does this implement/fix? Explain your changes.
Add an histogram to the calibration plot. For now, works only when only one classifier per plot (no staking).

#### Any other comments?
Work in progress

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

aperezlebel committed:
Add support for custom bins to calibration_curve

aperezlebel committed:
Update CalibrationDisplay to plot hist on subaxis

(For now works for a single estimator only)

aperezlebel committed:
Add from_estimators

glemaitre commented on sklearn/calibration.py:
```diff
@@ -934,9 +954,10 @@ def calibration_curve(
             `predict_proba` positive class).
 
-    n_bins : int, default=5
+    n_bins : int or sequence, default=5
```
```suggestion
    n_bins : int or array-like, default=5
```

glemaitre commented on sklearn/calibration.py:
```diff
@@ -934,9 +954,10 @@ def calibration_curve(
-    n_bins : int, default=5
+    n_bins : int or sequence, default=5
         Number of bins to discretize the [0, 1] interval. A bigger number
-        requires more data. Bins with no samples (i.e. without
+        requires more data. A sequence of bins can also be given.
+        Bins with no samples (i.e. without
         corresponding values in `y_prob`) will not be returned, thus the
         returned arrays may have less than `n_bins` values.
```
I think that I would rewrite it as:

```
Define the discretization applied to `y_prob`, ranging in [0, 1].

- if an integer is provided, the discretization depends on the
  `strategy` parameter.
- if an array-like is provided, the `strategy` parameter is overlooked
  and the array is used as bin edges directly.

A bigger requested number of bins require more data. Bins with no samples
(i.e. without corresponding values in `y_prob`) will not be returned, thus
the returned arrays may have less than `n_bins` values (or `len(n_bins)`
if `n_bins` is an array-like).
```

glemaitre commented on sklearn/calibration.py:
```diff
@@ -892,6 +893,25 @@ def predict(self, T):
         return expit(-(self.a_ * T + self.b_))
 
 
+def bins_from_strategy(n_bins, strategy, y_prob=None):
```
```suggestion
def bins_from_strategy(n_bins, strategy, y_prob=None):
    """Define the bin edges based on the strategy.
    
    If `n_bins` is already an `array-like`, it is used as-is by
    converting it to a `ndarray`.
    """
```

glemaitre commented on sklearn/calibration.py:
```diff
@@ -948,6 +969,8 @@ def calibration_curve(
         quantile
             The bins have the same number of samples and depend on `y_prob`.
 
+        Ignored if n_bins is an array of bins.
```
```suggestion
        Ignored if `n_bins` is an array-like containing the bin edges.
```

glemaitre commented on sklearn/calibration.py:
```diff
@@ -1158,17 +1185,47 @@ def plot(self, *, ax=None, name=None, ref_line=True, **kwargs):
+
+        if plot_bins:
+            for x in self.bins:
+                ax.axvline(x, lw=0.5, ls="--", color="grey", zorder=-1)
```
I assume that we should probably store those values as well if we want to be consistent.
The same for the grid.

aperezlebel replied:
> rename plot_bins to grid

Good idea! I made the change.

> store those values


Do you mean storing the variable grid ?

glemaitre commented on sklearn/calibration.py:
```diff
@@ -1295,9 +1354,150 @@ def from_estimator(
         )
 
+    @classmethod
+    def from_estimators(
```
I will wait for reviewing this part for the moment.

aperezlebel committed:
Update

aperezlebel commented:
Thank you @glemaitre for the review!
Working on it! @ogrisel

aperezlebel commented on sklearn/calibration.py:
```diff
@@ -892,6 +893,25 @@ def predict(self, T):
 
 
+def bins_from_strategy(n_bins, strategy, y_prob=None):
+    try:
+        if strategy == "quantile":  # Determine bin edges by distribution of data
+            quantiles = np.linspace(0, 1, n_bins + 1)
+            bins = np.percentile(y_prob, quantiles * 100)
+        elif strategy == "uniform":
+            bins = np.linspace(0.0, 1.0, n_bins + 1)
+        else:
+            raise ValueError(
+                "Invalid entry to 'strategy' input. Strategy "
+                "must be either 'quantile' or 'uniform'."
+            )
+
+    except TypeError:  # n_bins is not a scalar
+        bins = n_bins  # assumed to be an array-like of bins
+
+    return bins
```
Shouldn't we keep the error message when strategy is not "uniform" or "quantile"?

aperezlebel committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

aperezlebel committed:
Fix histogram and plot uniform bins

aperezlebel committed:
Change plot_bins to grid keyword

aperezlebel committed:
Update calibration example with new implementation

aperezlebel committed:
Plot hist labels in CalibrationDisplay

aperezlebel committed:
Update doc for n_bins in calibration_curve

aperezlebel committed:
Compute bins_hist on object creation

aperezlebel commented on sklearn/calibration.py:
```diff
@@ -1158,17 +1185,47 @@ def plot(self, *, ax=None, name=None, ref_line=True, **kwargs):
+
+        ax_hist.hist(
+            self.y_prob,
+            bins=bins_from_strategy(len(self.bins), strategy="uniform"),
```
I added a `bins_hist` attribute that is computed in `from_predictions` before object creation.

aperezlebel committed:
Add option to plot labels on histogram

aperezlebel committed:
Clean up and update docstrings

aperezlebel commented on sklearn/calibration.py:
```diff
@@ -1158,17 +1185,47 @@ def plot(self, *, ax=None, name=None, ref_line=True, **kwargs):
+            density=True,
+            histtype="step",
+            color=self.line_.get_color(),
+        )
```
I added a `bin_label` option to plot the histogram bar ratios.

aperezlebel committed:
Update Probability Calibration curves example

aperezlebel commented:
I addressed the comments (@glemaitre) and updated the doc examples ("Comparison of Calibration of Classifiers" and "Probability Calibration curves") with the new implementation (@ogrisel).

Some points remain to address:
* Add an option to not plot histogram at all. For now when `ax_hist=None`, an axis is created from `ax` to plot the histogram.
* Plotting several classifiers on the same axis using `from_estimators` is not functional as histograms are illegible.
* If we keep `from_estimators`, I think we should add a counterpart with predictions to provide the same functionality when users don't have estimators.

aperezlebel committed:
Fix failing tests in test_calibration

aperezlebel committed:
Fix failing docstring tests

aperezlebel committed:
Add entry to changelog

aperezlebel committed:
Fix sphinx warning

aperezlebel committed:
Fix failing test when bins_hist is empty

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/aperezlebel/23748-2

glemaitre commented on sklearn/calibration.py:
```diff
@@ -1310,9 +1484,150 @@ def from_estimator(
+        grid=True,
+        **kwargs,
+    ):
+        """Plot calibration curve using a binary classifier and data.
```
I have to check the discussion to recall why we want this function. Looking at the direction that I would like to take for the PDP (1-call -> 1 plot), I would be tented to not add this method but I have to check what is the benefit.

If I am not wrong, we don't have any example that demonstrate the usage of this method, isn't it?

glemaitre commented on sklearn/calibration.py:
```diff
@@ -1310,9 +1484,150 @@ def from_estimator(
+            for i in range(len(estimators))
+        ]
+
+        for i, (estimator, name) in enumerate(zip(estimators, names)):
```
Looking at the core of the method, I would say that having the example of making the call in a `for` loop would be enough.
